File: app/src/main/java/at/bitfire/davdroid/ui/setup/DetectConfigurationFragment.java
/*
 * Copyright © 2013 – 2016 Ricki Hirner (bitfire web engineering).
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the GNU Public License v3.0
 * which accompanies this distribution, and is available at
 * http://www.gnu.org/licenses/gpl.html
 */

package at.bitfire.davdroid.ui.setup;

import android.app.Dialog;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.AsyncTaskLoader;
import android.support.v4.content.Loader;
import android.support.v7.app.AlertDialog;

import at.bitfire.davdroid.R;
import at.bitfire.davdroid.ui.DebugInfoActivity;
import at.bitfire.davdroid.ui.setup.DavResourceFinder.Configuration;

public class DetectConfigurationFragment extends DialogFragment implements LoaderManager.LoaderCallbacks<Configuration> {
    protected static final String ARG_LOGIN_CREDENTIALS = "credentials";

    public static DetectConfigurationFragment newInstance(LoginCredentials credentials) {
        DetectConfigurationFragment frag = new DetectConfigurationFragment();
        Bundle args = new Bundle(1);
        args.putParcelable(ARG_LOGIN_CREDENTIALS, credentials);
        frag.setArguments(args);
        return frag;
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        ProgressDialog dialog = new ProgressDialog(getActivity());
        dialog.setCanceledOnTouchOutside(false);
        setCancelable(false);

        dialog.setTitle(R.string.login_configuration_detection);
        dialog.setIndeterminate(true);
        dialog.setMessage(getString(R.string.login_querying_server));
        return dialog;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        getLoaderManager().initLoader(0, getArguments(), this);
    }

    @Override
    public Loader<Configuration> onCreateLoader(int id, Bundle args) {
        return new ServerConfigurationLoader(getContext(), (LoginCredentials)args.getParcelable(ARG_LOGIN_CREDENTIALS));
    }

    @Override
    public void onLoadFinished(Loader<Configuration> loader, Configuration data) {
        if (data.calDAV == null && data.cardDAV == null)
            // no service found: show error message
            getFragmentManager().beginTransaction()
                    .add(NothingDetectedFragment.newInstance(data.logs), null)
                    .commitAllowingStateLoss();
        else
            // service found: continue
            getFragmentManager().beginTransaction()
                    .replace(R.id.fragment, AccountDetailsFragment.newInstance(data))
                    .addToBackStack(null)
                    .commitAllowingStateLoss();

        dismissAllowingStateLoss();
    }

    @Override
    public void onLoaderReset(Loader<Configuration> loader) {
    }


    public static class NothingDetectedFragment extends DialogFragment {
        private static String KEY_LOGS = "logs";

        public static NothingDetectedFragment newInstance(String logs) {
            Bundle args = new Bundle();
            args.putString(KEY_LOGS, logs);
            NothingDetectedFragment fragment = new NothingDetectedFragment();
            fragment.setArguments(args);
            return fragment;
        }

        @Override
        public Dialog onCreateDialog(Bundle savedInstanceState) {
            return new AlertDialog.Builder(getActivity())
                    .setTitle(R.string.login_configuration_detection)
                    .setIcon(R.drawable.ic_error_dark)
                    .setMessage(R.string.login_no_caldav_carddav)
                    .setNeutralButton(R.string.login_view_logs, new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            Intent intent = new Intent(getActivity(), DebugInfoActivity.class);
                            intent.putExtra(DebugInfoActivity.KEY_LOGS, getArguments().getString(KEY_LOGS));
                            startActivity(intent);
                        }
                    })
                    .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            // dismiss
                        }
                    })
                    .create();
        }
    }

    static class ServerConfigurationLoader extends AsyncTaskLoader<Configuration> {
        final Context context;
        final LoginCredentials credentials;

        public ServerConfigurationLoader(Context context, LoginCredentials credentials) {
            super(context);
            this.context = context;
            this.credentials = credentials;
        }

        @Override
        protected void onStartLoading() {
            forceLoad();
        }

        @Override
        public Configuration loadInBackground() {
            DavResourceFinder finder = new DavResourceFinder(context, credentials);
            return finder.findInitialConfiguration();
        }
    }
}

File: app/src/main/kotlin/at/bitfire/davdroid/ui/setup/DetectConfigurationFragment.kt
/***************************************************************************************************
 * Copyright © All Contributors. See LICENSE and AUTHORS in the root directory for details.
 **************************************************************************************************/

package at.bitfire.davdroid.ui.setup

import android.app.Application
import android.app.Dialog
import android.os.Bundle
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.activity.compose.BackHandler
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material.LinearProgressIndicator
import androidx.compose.material.MaterialTheme
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.fragment.app.viewModels
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.MutableLiveData
import at.bitfire.davdroid.R
import at.bitfire.davdroid.db.Credentials
import at.bitfire.davdroid.log.Logger
import at.bitfire.davdroid.servicedetection.DavResourceFinder
import at.bitfire.davdroid.ui.AppTheme
import at.bitfire.davdroid.ui.DebugInfoActivity
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import kotlinx.coroutines.CoroutineName
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.plus
import kotlinx.coroutines.runInterruptible
import java.net.URI
import java.util.logging.Level

class DetectConfigurationFragment: Fragment() {

    private val loginModel by activityViewModels<LoginModel>()
    private val model by viewModels<DetectConfigurationModel>()


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val baseURI = loginModel.baseURI ?: return

        model.result.observe(this) { result ->
            // save result for next step
            loginModel.configuration = result

            // remove "Detecting configuration" fragment, it shouldn't come back
            parentFragmentManager.popBackStack()

            if (result.calDAV != null || result.cardDAV != null)
                parentFragmentManager.beginTransaction()
                        .replace(android.R.id.content, AccountDetailsFragment())
                        .addToBackStack(null)
                        .commit()
            else
                parentFragmentManager.beginTransaction()
                        .add(NothingDetectedFragment(), null)
                        .commit()
        }

        model.start(baseURI, loginModel.credentials)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?) =
        ComposeView(requireContext()).apply {
            setContent {
                AppTheme {
                    DetectConfigurationView()
                }

                // Cancel service detection only when back button is pressed
                BackHandler {
                    model.cancel()

                    parentFragmentManager.popBackStack()
                }
            }
        }


    class DetectConfigurationModel(application: Application): AndroidViewModel(application) {

        val scope = MainScope() + CoroutineName("DetectConfigurationModel")
        var result = MutableLiveData<DavResourceFinder.Configuration>()

        /**
         * Starts service detection in a global scope which is independent of the ViewModel scope so
         * that service detection won't be cancelled when the user for instance switches to another app.
         *
         * The service detection result will be posted into [result].
         */
        fun start(baseURI: URI, credentials: Credentials?) {
            scope.launch(Dispatchers.Default) {
                runInterruptible {
                    try {
                        DavResourceFinder(getApplication(), baseURI, credentials).use { finder ->
                            val configuration = finder.findInitialConfiguration()
                            result.postValue(configuration)
                        }
                    } catch(e: Exception) {
                        // This shouldn't happen; instead configuration should be empty
                        Logger.log.log(Level.WARNING, "Uncaught exception during service detection, shouldn't happen", e)
                    }
                }
            }
        }

        /**
         * Cancels a potentially running service detection.
         */
        fun cancel() {
            Logger.log.info("Aborting resource detection")
            try {
                scope.cancel()
            } catch (ignored: IllegalStateException) { }
        }

    }


    class NothingDetectedFragment: DialogFragment() {

        val model by activityViewModels<LoginModel>()

        override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
            var message = getString(R.string.login_no_caldav_carddav)
            if (model.configuration?.encountered401 == true)
                message += "\n\n" + getString(R.string.login_username_password_wrong)

            return MaterialAlertDialogBuilder(requireActivity())
                    .setTitle(R.string.login_configuration_detection)
                    .setIcon(R.drawable.ic_error)
                    .setMessage(message)
                    .setNeutralButton(R.string.login_view_logs) { _, _ ->
                        val intent = DebugInfoActivity.IntentBuilder(requireActivity())
                            .withLogs(model.configuration?.logs)
                            .build()
                        startActivity(intent)
                    }
                    .setPositiveButton(android.R.string.ok) { _, _ ->
                        // just dismiss
                    }
                    .create()
        }

    }

}


@Composable
fun DetectConfigurationView() {
    Column(Modifier
        .fillMaxWidth()
        .padding(8.dp)) {
        Text(
            stringResource(R.string.login_configuration_detection),
            style = MaterialTheme.typography.h5
        )
        LinearProgressIndicator(
            color = MaterialTheme.colors.secondary,
            modifier = Modifier
                .fillMaxWidth()
                .padding(top = 8.dp, bottom = 16.dp))
        Text(
            stringResource(R.string.login_querying_server),
            style = MaterialTheme.typography.body1
        )
    }
}

@Composable
@Preview
fun DetectConfigurationView_Preview() {
    DetectConfigurationView()
}